### ISSUE TITLE
Gramine-Direct Fuzzing: Crash reports as "Internal Memory Fault" when Gramine is build in release mode

### ISSUE THREAD

**Issue by anjalirai-intel** [gramineproject/gramine#1521]
### Description of the problem

Syzkaller has been modified in order to run with Gramine (https://github.com/JaewonHur/syzkaller). We were able to run it with Gramine v1.5 for Gramine-Direct

The attached crash reports as "Internal Memory Fault" when Gramine is build in release mode, does throw any asserts

### Steps to reproduce

1. Extract the zip file and cd to the drive
2. Build the Gramine(v1.5) in Release mode
3. make DEBUG=1 CRASH=6682fcdd5ac4cf13
4. gramine-direct crash
[crash-68b52d938b74b3a5_release.zip](https://github.com/gramineproject/gramine/files/12453422/crash-68b52d938b74b3a5_release.zip)

### Actual results

```
(libos_init.c:505:libos_init) [P2:T2:crash-68b52d938b74b3a5] debug: LibOS initialized
(libos_parser.c:1628:buf_write_all) [P2:T2:crash-68b52d938b74b3a5] trace: ---- set_robust_list(0x1c215270ea20, 0x18) = 0x0
(libos_parser.c:1628:buf_write_all) [P1:T1:crash-68b52d938b74b3a5] trace: ---- return from clone(...) = 0x2
(libos_parser.c:1628:buf_write_all) [P1:T1:crash-68b52d938b74b3a5] trace: ---- wait4(-1, 0x1c2152964d7c, |0x40000000, 0) ...
(libos_parser.c:1628:buf_write_all) [P2:T2:crash-68b52d938b74b3a5] trace: ---- write(1, 0x1c2152921004, 0x12) ...
executing program
(libos_parser.c:1628:buf_write_all) [P2:T2:crash-68b52d938b74b3a5] trace: ---- return from write(...) = 0x12
(libos_parser.c:1628:buf_write_all) [P2:T2:crash-68b52d938b74b3a5] trace: ---- write(2, 0x1c2152921017, 0xa) ...
### start
(libos_parser.c:1628:buf_write_all) [P2:T2:crash-68b52d938b74b3a5] trace: ---- return from write(...) = 0xa
(libos_signal.c:311:internal_fault) [P2:T2:crash-68b52d938b74b3a5] error: Internal memory fault at 0x1c2152965088 (0x7ffffff62903, VMID = 2, TID = 2)
(libos_ipc_child.c:21:ipc_child_disconnect_callback) [P1:libos] error: Child process (vmid: 0x2) got disconnected
(libos_parser.c:1628:buf_write_all) [P1:T1:crash-68b52d938b74b3a5] trace: ---- return from wait4(...) = -512
(libos_parser.c:1628:buf_write_all) [P1:T1:crash-68b52d938b74b3a5] trace: ---- wait4(-1, 0x1c2152964d7c, |0x40000000, 0) ...
(libos_parser.c:1628:buf_write_all) [P1:T1:crash-68b52d938b74b3a5] trace: ---- return from wait4(...) = 0x2
(libos_exit.c:212:libos_syscall_exit_group) [P1:T1:crash-68b52d938b74b3a5] debug: ---- exit_group (returning 0)
(libos_fs_lock.c:785:file_lock_clear_pid) [P1:T1:crash-68b52d938b74b3a5] debug: clearing file (POSIX) locks for pid 1
(libos_sync_client.c:331:shutdown_sync_client) [P1:T1:crash-68b52d938b74b3a5] debug: sync client shutdown: closing handles
(libos_sync_client.c:346:shutdown_sync_client) [P1:T1:crash-68b52d938b74b3a5] debug: sync client shutdown: waiting for confirmation
(libos_sync_client.c:359:shutdown_sync_client) [P1:T1:crash-68b52d938b74b3a5] debug: sync client shutdown: finished
(libos_ipc_worker.c:285:ipc_worker_main) [P1:libos] debug: IPC worker: exiting worker thread
(libos_exit.c:58:libos_clean_and_exit) [P1:T1:crash-68b52d938b74b3a5] debug: process 1 exited with status 0
```

### Gramine commit hash

0b552696d4f402c291c8c2cce525e7117d06d397

**Comment by llly** [comment]
It's caused by an assert.
https://github.com/gramineproject/gramine/blob/e4e7ab76f121679bbef2558c1b36a2af5fed9160/libos/src/arch/x86_64/libos_context.c#L113
I run with debug gramine-direct and add a log to print xstate address.
```
(libos_context.c:113:is_xstate_extended) [P2:T2:crash-68b52d938b74b3a5] debug: xstate at 0x35937b3baea8
(libos_context.c:114:is_xstate_extended) [P2:T2:crash-68b52d938b74b3a5] assert failed ../libos/src/arch/x86_64/libos_context.c:114 IS_ALIGNED_PTR(xstate, LIBOS_XSTATE_ALIGN)
```

**Comment by anjalirai-intel** [comment]
Yes, it is same, but why it does not assert in release mode

**Comment by jinengandhi-intel** [comment]
Just to expand on what Anjali said, #1515 is for the test failing with an `assert` when Gramine is built in debug mode and this issue for the same test failing with an `Internal memory fault` when Gramine is built in release mode.

**Comment by dimakuv** [comment]
> Yes, it is same, but why it does not assert in release mode

Asserts are removed in the release mode. That's what "asserts" mean -- they are helpers for developers that only exist in debug mode of the application. See e.g. https://softwareengineering.stackexchange.com/questions/364145/should-there-be-assertions-in-release-builds

**Comment by dimakuv** [comment]
Since this issue has the same root cause as #1515, I will close this as a duplicate and mention in #1515 how it fails in release mode.

**Comment by jinengandhi-intel** [comment]
Program failing with an `internal memory fault` in release mode, isn't this a bigger problem than the `assert` in debug mode? I am okay we closing 1 issue but just want to clarify that we plan to look at both issues?

**Comment by dimakuv** [comment]
> Program failing with an `internal memory fault` in release mode, isn't this a bigger problem than the `assert` in debug mode? I am okay we closing 1 issue but just want to clarify that we plan to look at both issues?

Yes, it is a bigger problem. But the root cause for both problems is the same -- thus, from a Gramine developer perspective, there is no reason to have two separate GitHub issues that describe the same problem. In other words, when a Gramine dev fixes this debug-mode-related issue, he will automatically fix the release-mode-related issue (and the other way around).

I closed the other GitHub issue purely to decrease the number of opened GitHub issues -- we already get lost in them... And whoever will fix this particular issue, will need to check that the 1515 case was also fixed.
